### PR TITLE
[SYCLomatic] Wrap sycl::buffer_allocator into dpct::detail::__buffer_allocator

### PR DESCRIPTION
**Pull request by yihwang-nv**
[SYCL2020] Implement sycl::buffer_allocator which takes a template argument

Signed-off-by: Wang, Yihan <yihan.wang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc [hunk 1]
```diff
@@ -49,6 +49,15 @@
 // device_new, free_device, device_delete
 // DPCT_LABEL_END
 namespace dpct {
+namespace detail {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc:52]:
```suggestion
namespace detail {
// DPCT_LABEL_BEGIN|__buffer_allocator|dpct::detail
// DPCT_DEPENDENCY_EMPTY
// DPCT_CODE
template <typename _DataT>
using __buffer_allocator =
#if __LIBSYCL_VERSION >= 50707
    sycl::buffer_allocator<_DataT>;
#else
    sycl::buffer_allocator;
#endif
// DPCT_LABEL_END
} // namespace detail
```

Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc:52]:
```suggestion
namespace detail {
// DPCT_LABEL_BEGIN|__buffer_allocator|dpct::detail
// DPCT_DEPENDENCY_EMPTY
// DPCT_CODE
template <typename _DataT>
using __buffer_allocator =
#if __LIBSYCL_VERSION >= 50707
    sycl::buffer_allocator<_DataT>;
#else
    sycl::buffer_allocator;
#endif
// DPCT_LABEL_END
} // namespace detail
```

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc [hunk 2]
```diff
@@ -57,7 +66,7 @@ namespace dpct {
 // DPCT_CODE
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc:71]: Add dependency
```
DplExtrasMemory|__buffer_allocator
```

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc [hunk 3]
```diff
@@ -524,7 +533,7 @@ public:
...
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc:539]: Add dependency
```
DplExtrasMemory|__buffer_allocator
```

File: clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc [hunk 4]
```diff
@@ -409,10 +409,10 @@ public:
 
 #else
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc:412]: Add dependency
```
DplExtrasMemory|__buffer_allocator
```

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc (hunk 1 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc:52]: +1

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc (hunk 2 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc:71]: +1

File: clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc (hunk 3 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpl_extras/memory.h.inc:539]: +1

File: clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc (hunk 4 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc:412]: +1

**Review by zhiweij1 — APPROVED** [review]

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**